[FLINK-32310][table] Support enhanced show functions syntax (#22903)
apache/flink: docs/content/docs/dev/table/sql/show.md

## SHOW FUNCTIONS

```sql
SHOW [USER] FUNCTIONS [ ( FROM | IN ) [catalog_name.]database_name ] [ [NOT] (LIKE | ILIKE) <sql_like_pattern> ]	
```

> Show all functions including system functions and user-defined functions for an optionally specified database. If no database is specified then the functions are returned from the current database. Additionally, the output of this statement may be filtered by an optional matching pattern.

Reviewer: ```suggestion
Show all functions including system functions and user-defined functions for an optionally specified database. If no database is specified then the functions are returned from the current database. Additionally, a `sql_like_pattern` can be used to filter the functions.
```

**USER**

> Show only user-defined functions for an optionally specified database. If no database is specified then the functions are returned from the current database. Additionally, the output of this statement may be filtered by an optional matching pattern.

Reviewer: ```suggestion
Show only user-defined functions for an optionally specified database. If no database is specified then the functions are returned from the current database. Additionally, a `sql_like_pattern` can be used to filter the functions.
```

**LIKE**
Show all tables with given table name and optional `LIKE` clause, whose name is whether similar to the `<sql_like_pattern>`.

The syntax of sql pattern in `LIKE` clause is the same as that of `MySQL` dialect.

> * `%` matches any number of characters, even zero characters, `\%` matches one `%` character.

Reviewer: ```suggestion
* `%` matches any number of characters, even zero characters, and `\%` matches one `%` character.
```

* `_` matches exactly one character, `\_` matches one `_` character.

**ILIKE**

> The same behavior as `LIKE` but sql pattern is case-insensitive.

Reviewer: ```suggestion
The same behavior as `LIKE` but the SQL pattern is case-insensitive.
```
